app/scripts/lib/multichain-api/provider-authorize/handler.js

```diff
@@ -0,0 +1,213 @@
+  assertScopesSupported,
+} from '../scope';
+import {
+  Caip25CaveatType,
+  Caip25EndowmentPermissionName,
+} from '../caip25permissions';
+import { assignAccountsToScopes, validateAndUpsertEip3085 } from './helpers';
+
+const getAccountsFromPermission = (permission) => {
+  return permission.eth_accounts.caveats.find(
+    (caveat) => caveat.type === 'restrictReturnedAccounts',
+  )?.value;
+};
+
+// TODO:
+// Unless the dapp is known and trusted, give generic error messages for
+// - the user denies consent for exposing accounts that match the requested and approved chains,
+// - the user denies consent for requested methods,
+// - the user denies all requested or any required scope objects,
+// - the wallet cannot support all requested or any required scope objects,
+// - the requested chains are not supported by the wallet, or
+// - the requested methods are not supported by the wallet
+// return
+//     "code": 0,
+//     "message": "Unknown error"
+
+// TODO:
+// When user disapproves accepting calls with the request methods
+//   code = 5001
+//   message = "User disapproved requested methods"
+// When user disapproves accepting calls with the request notifications
+//   code = 5002
+//   message = "User disapproved requested notifications"
+
+export async function providerAuthorizeHandler(req, res, _next, end, hooks) {
+  // TODO: Does this handler need a rate limiter/lock like the one in eth_requestAccounts?
+
+  const {
+    origin,
+    params: {
+      requiredScopes,
+      optionalScopes,
+      sessionProperties,
+      scopedProperties,
+      ...restParams
+    },
+  } = req;
+
+  const { findNetworkClientIdByChainId } = hooks;
+
+  if (Object.keys(restParams).length !== 0) {
+    return end(
+      new EthereumRpcError(
+        5301,
+        'Session Properties can only be optional and global',
+      ),
+    );
+  }
+
+  const sessionId = '0xdeadbeef';
+
+  if (sessionProperties && Object.keys(sessionProperties).length === 0) {
+    return end(
+      new EthereumRpcError(5300, 'Invalid Session Properties requested'),
+    );
+  }
+
+  const networkClientIdsAdded = [];
+
+  try {
+    const { flattenedRequiredScopes, flattenedOptionalScopes } = processScopes(
+      requiredScopes,
+      optionalScopes,
+    );
+
+    const validScopedProperties = processScopedProperties(
+      flattenedRequiredScopes,
+      flattenedOptionalScopes,
+      scopedProperties,
+    );
+
+    const existsNetworkClientForChainId = (chainId) => {
+      try {
+        findNetworkClientIdByChainId(chainId);
+        return true;
+      } catch (err) {
+        return false;
+      }
+    };
+
+    const existsEip3085ForChainId = (chainId) => {
+      const scopeString = `eip155:${parseInt(chainId, 16)}`;
+      return validScopedProperties?.[scopeString]?.eip3085;
+    };
+
+    const {
+      supportedScopes: supportedRequiredScopes,
+      supportableScopes: supportableRequiredScopes,
+      unsupportedScopes: unsupportedRequiredScopes,
```

Comment:
total nit
```suggestion
      unsupportableScopes: unsupportableRequiredScopes,
```
and so on

Reply:
https://github.com/MetaMask/metamask-extension/pull/25873/commits/a0a4d0a6056a9e7fddba855b445ed08b64f1ad10